src/main/scala/pipelined/cpu.scala

```diff
@@ -128,179 +141,186 @@ class PipelinedCPU(implicit val conf: CPUConfig) extends BaseCPU {
+  val rs2 = if_id.io.data.instruction(24,20)
 
   // Send input from this stage to hazard detection unit
   hazard.io.rs1 := rs1
   hazard.io.rs2 := rs2
 
   // Send opcode to control
-  control.io.opcode := if_id.instruction(6,0)
+  control.io.opcode := if_id.io.data.instruction(6,0)
 
   // Send register numbers to the register file
   registers.io.readreg1 := rs1
   registers.io.readreg2 := rs2
 
   // Send the instruction to the immediate generator
-  immGen.io.instruction := if_id.instruction
+  immGen.io.instruction := if_id.io.data.instruction
 
+  // Data supplied to id_ex register is always valid every cycle
+  id_ex.io.valid := true.B
+  // Don't need to flush nor bubble the data in this register
+  id_ex.io.flush := false.B
+  id_ex.io.bubble := false.B
   // FIll the id_ex register
-  id_ex.writereg   := if_id.instruction(11,7)
-  id_ex.rs1        := rs1
-  id_ex.rs2        := rs2
-  id_ex.funct7     := if_id.instruction(31,25)
-  id_ex.funct3     := if_id.instruction(14,12)
-  id_ex.imm        := immGen.io.sextImm
-  id_ex.readdata2  := registers.io.readdata2
-  id_ex.readdata1  := registers.io.readdata1
-  id_ex.pc         := if_id.pc
-  id_ex.pcplusfour := if_id.pcplusfour
-
+  id_ex.io.in.writereg   := if_id.io.data.instruction(11,7)
+  id_ex.io.in.rs1        := rs1
+  id_ex.io.in.rs2        := rs2
+  id_ex.io.in.funct7     := if_id.io.data.instruction(31,25)
+  id_ex.io.in.funct3     := if_id.io.data.instruction(14,12)
+  id_ex.io.in.imm        := immGen.io.sextImm
+  id_ex.io.in.readdata2  := registers.io.readdata2
+  id_ex.io.in.readdata1  := registers.io.readdata1
+  id_ex.io.in.pc         := if_id.io.data.pc
+  id_ex.io.in.pcplusfour := if_id.io.data.pcplusfour
+
+  // Data supplied to id_ex_ctrl register is always valid every cycle
+  id_ex_ctrl.io.valid := true.B
+  // Don't need to bubble the data in this register
+  id_ex_ctrl.io.bubble := false.B
   // Set the execution control signals
-  id_ex.excontrol.add       := control.io.add
-  id_ex.excontrol.immediate := control.io.immediate
-  id_ex.excontrol.alusrc1   := control.io.alusrc1
-  id_ex.excontrol.branch    := control.io.branch
-  id_ex.excontrol.jump      := control.io.jump
+  id_ex_ctrl.io.in.ex.add       := control.io.add
+  id_ex_ctrl.io.in.ex.immediate := control.io.immediate
+  id_ex_ctrl.io.in.ex.alusrc1   := control.io.alusrc1
+  id_ex_ctrl.io.in.ex.branch    := control.io.branch
+  id_ex_ctrl.io.in.ex.jump      := control.io.jump
 
   // Set the memory control signals
-  id_ex.mcontrol.memread  := control.io.memread
-  id_ex.mcontrol.memwrite := control.io.memwrite
-  id_ex.mcontrol.maskmode := if_id.instruction(13,12)
-  id_ex.mcontrol.sext     := ~if_id.instruction(14)
+  id_ex_ctrl.io.in.m.memread  := control.io.memread
+  id_ex_ctrl.io.in.m.memwrite := control.io.memwrite
+  id_ex_ctrl.io.in.m.maskmode := if_id.io.data.instruction(13,12)
+  id_ex_ctrl.io.in.m.sext     := ~if_id.io.data.instruction(14)
+  id_ex_ctrl.io.in.m.taken    := false.B
```

Comment:
Was this a bug before? Did we forget to set the taken bit?

Reply:
Yes, but taken in the mcontrol block of the idex register isn't used anywhere as far as I can tell. I only have to set it here because of elaboration errors.

Reply:
Gotcha. :)